xcat-inventory/xcclient/inventory/structurediff.py

```diff
@@ -0,0 +1,176 @@
+                        if '}' in tmp_str:
+                            bracket -= 1
+                            if not bracket:
+                                break
+                        tmp_str = flag + tmp_str[4:]
+                        new_list.append(tmp_str)
+                else:
+                    tmp_str = tmp_str.replace('%s: ' % diff_flag, '')
+                    tmp_str = flag + tmp_str
+                    new_list.append(tmp_str)
+        
+            else:
+                new_list.append(tmp_str)
+
+        return ('\n'.join(new_list))
+
+    def deal_with_diff_dict(self, result_dict):
+        diff_dict = {}
+        for key, value in result_dict.items():
+            for change in value:
+                mychange = {}
+                if 'added' in key:
+                    if 'iterable' in key:
+                        path = self._get_path_as_list(change.up.path())
+                        extra = path.pop()
+                        mychange = {extra: ['+diff: %s' % change.t2]}
+                    else:
+                        path= self._get_path_as_list(change.path())
+                        extra = path.pop()
+                        mychange = { '+diff': {extra: change.t2}}
+                elif 'removed' in key: 
+                    if 'iterable' in key:
+                        path = self._get_path_as_list(change.up.path())
+                        extra = path.pop()
+                        mychange = {extra: ['-diff: %s' % change.t1]}
+                    else:
+                        path = self._get_path_as_list(change.path())
+                        extra = path.pop()
+                        mychange = {'-diff': {extra: change.t1}}
+                elif 'changed' in key:
+                    path = self._get_path_as_list(change.path())
+                    extra = path.pop()
+                    mychange = {'-diff': {extra: change.t1}, '+diff': {extra: change.t2}}
+
+                while len(path) > 0:
+                    key_str = path.pop()
+                    mychange = {key_str: mychange}
+
+                for change_key in mychange:
+                    diff_dict = self._update_dict(diff_dict, {change_key: mychange[change_key]}) 
+        return diff_dict
+
+    def get_diff_string(self, format_type, diff_dict):
+        if format_type == 'json':
+            diff_json = json.dumps(diff_dict, indent=4, separators=(',', ': '))
+            return (self._format_json(diff_json))
+        else:
+            diff_yaml = yaml.safe_dump(diff_dict, default_flow_style=False,allow_unicode=True)
+            return (self._format_yaml(diff_yaml))
+
+
+class StructureDiff(object):
+
+    def __init__(self):
+        pass
+
+    def _get_deepdiff(self, obj1, obj2):
+        self.diff = deepdiff.DeepDiff(obj1,obj2,ignore_order=True,report_repetition=False,exclude_paths='',significant_digits=None,view='tree',verbose_level=1)
+
+    def repet(self, diff_dict, fmt):
```

Comment:
should better be `rept`,there should be description in the comments before each method